docs/goals_and_soa.md

```diff
@@ -0,0 +1,392 @@
+(synced-by-construction), (secure-by-construction)
+and (declarative-spec).
+
+@todo #3: expand docs on Atlas
+
+## Testing tools
+
+### Tweag's cooked-validators
+
+Project definetly covers goal (production-ready),
+because it was sucessfully used in real-world audits.
+@todo #3: provide examples
+but only partially covers
+(emulate-anything) and (declarative-spec) goals.
+
+While it, surely, can do transaction emulation,
+it, to best of our knoweledge, does not have monad instance
+for real L1 blockchain. So you still cannot reuse same code
+across tests and real blockchain Tx construction.
+
+`TxSkel` datatype is similar to design we are implementing,
+and other parts API have a lot of cool DX decisions.
+But `cooked-validators` lack script transition model,
+so it cannot provide declarative Tx submission error handling
+we aim to get, because it does not have information to decide,
+if specific Tx input is critical (like auth token),
+or can be re-selected again (like coin-selected ADA for payment).
+
+Having declarative transition model is even
+more important for mutation testing purposes.
+`cooked-validators` gives Cardan Tx level API,
+for constructoion of mutations.
+That means, that you should select, write and evaluate
+them manually, depending on logic of your script.
+This lead to logic coupling between spec, L1 code and test code.
+Such coupling and overall manual construction
+complicates spec validation and might lead
+to attack vectors missed by mistake.
+
+Our API presumes that means of on-script validation,
+in our case behavour equivalence (aka bi-similarity)
+of delaratime CEM state-machine and real on-chain script
+is verified and validated semi-automatically.
+Also we plan to make some kind of vulerabilities covered by
+`cooked-validators` imposible by constuction, specifically modules:
+`ValidityRange`, `DoubleSat`, `AddInputsAndOutputs`
+and `OutPermutations`.
+
+Another important feature is support for LTL logic formulas,
+which are known language for specifying state transitions in BMC.
+It seems like this feature is subsumed by
+`quickcheck-dynamic` based model checking specification
+we are going to provide.
+
+### Hydra Mutations
+
+Hydra mutations are similar to cooked-validators in their design
+and have similar drawback. Thus they do not cover (synced-by-construction), (secure-by-construction)
+and (declarative-spec) goals.
+
+https://abailly.github.io/posts/mutation-testing.html
+
+@todo: examples of test duplication and existing bugs
+
+### tasty-plutus
+
+Only cover (emulate-anything) goal, does not aim to cover others.
+Is deprecated as well.
+But it has some interesting low-level features,
+like `plutus-size-check`, which code we would probably steal.
+
+### General Haskell verification tools
+
+We plan to use first two tools in our implementation.
+They both are tested in real prjects and documented well.
+
+Other tools are not applicable to our project.
+
+1. quickehck-dynamic
+2. SMV
```

Comment:
What's SMV?

Reply:
This is very usable Haskell SMT interface. It provides both DSL for SMT constraints encoding and solving and soe more complicated static analysis DSL examples on top of that.